analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs

```diff
@@ -567,5 +569,97 @@ public void Main<T>() where T : new()
             validator.ValidateContainsOperation(OperationKind.TypeParameterObjectCreation);
             validator.ValidateTag("Value", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
         }
+
+        [TestMethod]
+        public void Invocation_SetsNotNullOnInstance_CS()
+        {
+            const string code = @"
+public class Sample
+{
+    public void Main(Sample instanceArg, Sample extensionArg)
+    {
+        var preserve = true;
+        Tag(""BeforeInstance"", instanceArg);
+        Tag(""BeforeExtension"", extensionArg);
+        Tag(""BeforePreserve"", preserve);
+
+        instanceArg.InstanceMethod();
+        extensionArg.ExtensionMethod();
+        UntrackedSymbol().InstanceMethod(); // Is not invoked on any symbol, should not fail
+        preserve.ExtensionMethod();
+        preserve.ToString();
+
+        Tag(""AfterInstance"", instanceArg);
+        Tag(""AfterExtension"", extensionArg);
+        Tag(""AfterPreserve"", preserve);
+    }
+
+    private void InstanceMethod() { }
+    private void Tag(string name, object arg) { }
+    private Sample UntrackedSymbol() => this;
+}
+
+public static class Extensions
+{
+    public static void ExtensionMethod(this Sample s) { }
+    public static void ExtensionMethod(this bool b) { }
+}";
+            var validator = new SETestContext(code, AnalyzerLanguage.CSharp, Array.Empty<SymbolicCheck>()).Validator;
+            validator.ValidateContainsOperation(OperationKind.Invocation);
+            validator.ValidateTag("BeforeInstance", x => x.Should().BeNull());
```

Comment:
We have tons of validations on `SymbolicValue`, `SymbolicContext` and other related types. We should consider adding custom validations to increase readability and remove duplication. Maybe a small suite of assertions like this:
```cs
validator.ValidateTag("BeforeInstance", x => x.Should().HaveNoConstraints());
validator.ValidateTag("AfterInstance", x => x.Should().HaveConstraint(ObjectConstraint.NotNull, "Instance method should set NotNull constraint."));
validator.ValidateTag("AfterInstance", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull, "Instance method should set NotNull constraint."));
validator.ValidateTag("AfterInstance", x => x.Should().HaveAtLeastConstraint(ObjectConstraint.NotNull, SomeOtherConstraint.Value);
```
See [fluentassertions extensibility](https://fluentassertions.com/extensibility/). A sample can also be found in #5880: https://github.com/SonarSource/sonar-dotnet/blob/a05abc4c2cddee7823e85b35307f27f788c69312/analyzers/tests/SonarAnalyzer.UnitTest/Extensions/BlockSyntaxExtensionsTests.cs#L273-L314

What do you think?

Reply:
We can consider that. It should look like
```
validator.TagValue("BeforeInstance").Should().HaveNoConstraints();
```